sensing/src/main/java/com/google/android/sensing/upload/SensingSynchronizer.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.sensing.upload
+
+import com.google.android.sensing.SensingEngine
+import com.google.android.sensing.model.RequestStatus
+import com.google.android.sensing.model.UploadResult
+import kotlinx.coroutines.awaitCancellation
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+
+class SensingSynchronizer(
+  private val sensingEngine: SensingEngine,
+  private val uploader: Uploader,
+  private val uploadResultProcessor: UploadResultProcessor
+) {
+
+  /**
+   * Sync Upload API. It fetches [UploadRequest]s, invokes [uploader], collects [UploadResult]s
+   * processes them using [UploadResult]s by [uploadResultProcessor]. At each stage it emits
+   * [SyncUploadState]s.
+   *
+   * TODO Upload until there is no upload requests in the database. Use UploadRequestPublisher
+   *
+   * TODO Persist the terminal SyncUploadState due to this
+   * [issue](https://github.com/google/android-fhir/issues/2119).
+   */
+  suspend fun synchronize(): Flow<SyncUploadState> = flow {
+    val uploadRequestList =
+      (sensingEngine.listUploadRequest(status = RequestStatus.UPLOADING) +
+        sensingEngine.listUploadRequest(status = RequestStatus.PENDING))
+
+    var lastSyncUploadState = SyncUploadState.Started(uploadRequestList.size) as SyncUploadState
+    emit(lastSyncUploadState)
+    if (uploadRequestList.isNotEmpty()) {
```

Comment:
What if the `uploadRequestList.isNotEmpty()` is empty? Should we just complete it immediately?

Reply:
Done!

Added lines
```
else { // no requests to process
      emit(SyncUploadState.Completed(0))
    }
```